### PR TITLE
Prevented buffer overflow when policy variable names are longer than 1024 bytes

### PR DESCRIPTION
**Pull request by olehermanse**
The function which checks if a variable name is valid, CheckParseVariable, was wrong in so many ways:

1. No unit test
2. Very hard to read
3. Gives wrong error messages (Always "reserved name", even when that is not the issue)
4. Stack buffer overflow for variables longer than 1024 bytes
5. For some edge cases like, `"a."`, the result is random (dependent on uninitialized memory)
6. It does a lot of passes over the same data (strlen, strlen, strchr, sscanf)
7. It's also just wrong, `".a"` is invalid, so is `".a[b]"`, but `".a[b.c]"` is valid when it clearly shouldn't be

(I'll fix 7. in a separate PR).

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2666/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2666/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2666/

**Comment by olehermanse** [comment]
@cf-bottom can you test this in jenkins (I added a unit test).

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2673/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2673/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2673/

**Comment by olehermanse** [comment]
@cf-bottom Now I'd like another jenkins build, since I refactored the code.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2674/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2674/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2674/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2675/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2675/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2675/

**Comment by olehermanse** [comment]
@cf-bottom one more jenkins build, I fixed the buffer overflow now.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2676/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2676/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2676/

**Comment by olehermanse** [comment]
@cf-bottom Now I changed the error message, another jenkins build! 🚀

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2678/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2678/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2678/

File: tests/unit/syntax_test.c [hunk 1]
```diff
@@ -109,6 +109,54 @@ static void test_typecheck_null_rval(void)
     assert_int_equal(SYNTAX_TYPE_MATCH_ERROR_GOT_NULL, err);
 }
 
+static void test_check_parse_variable_name(void)
+{
+    // Test was added after function
+    // It shows what it actually does (not what was intended)
+
+    // Allowed variable "names":
+    assert_true(CheckParseVariableName("a") == true);
```
Comment (vpodzime) [inline tests/unit/syntax_test.c]: I don't like this `assert_true(SOMETHING == true)`, it looks to me like `if (SOME_BOOL == true)`. And the same applies to to comparisons to `false`, I'd just use `assert_false()`.

File: libpromises/syntax.c [hunk 2]
```diff
@@ -889,9 +885,7 @@ bool CheckParseVariableName(const char *const name)
         if (count == 1)
         {
             // Check that there is something before and after first dot:
-            sscanf(name, "%[^.].%s", scopeid, vlval);
-
-            if (scopeid[0] == '\0' || vlval[0] == '\0')
+            if (name == first_dot || first_dot[1] == '\0')
```
Comment (vpodzime) [inline libpromises/syntax.c]: `name[0] == '.'` would be more understandable for me here.

File: libpromises/mod_common.c [hunk 3]
```diff
@@ -54,8 +54,8 @@ static const char *const POLICY_ERROR_VARS_CONSTRAINT_DUPLICATE_TYPE =
     "redefine with %s";
 static const char *const POLICY_ERROR_VARS_PROMISER_NUMERICAL =
     "Variable promises cannot have a purely numerical promiser (name)";
-static const char *const POLICY_ERROR_VARS_PROMISER_RESERVED =
-    "Variable promise is using a reserved name";
+static const char *const POLICY_ERROR_VARS_PROMISER_INVALID =
+    "Variable promise is using an invalid name (promiser)";
```
Comment (vpodzime) [inline libpromises/mod_common.c]: All the other errors say *promiser (name)*, I think we should be consistent here.

Comment (olehermanse) [reply libpromises/mod_common.c]: I'd like to change them all to be name (promiser) :) Focus on the easy terminology for new users.

File: tests/unit/syntax_test.c (hunk 1 above)
Comment (olehermanse) [reply tests/unit/syntax_test.c]: I guess it's possible that there is a difference, (non-zero vs equal to 1), but it makes sense to optimize for readability instead :) Fixed.

File: libpromises/syntax.c (hunk 2 above)
Comment (olehermanse) [reply libpromises/syntax.c]: Alright.

File: libpromises/mod_common.c (hunk 3 above)
Comment (olehermanse) [reply libpromises/mod_common.c]: Changed the other error messages to also be `name (promiser)`.

**Comment by olehermanse** [comment]
@cf-bottom jenkins, again :)

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2683/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2683/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2683/

**Comment by olehermanse** [comment]
@vpodzime I addressed your comments, please re-review.

**Comment by olehermanse** [comment]
Cherry-pick: https://github.com/cfengine/core/pull/3838
